libsolidity/ast/AST.h

```diff
@@ -597,12 +601,33 @@ class ContractDefinition: public Declaration, public StructurallyDocumented, pub
 	std::vector<ASTPointer<ASTNode>> m_subNodes;
 	ContractKind m_contractKind;
 	bool m_abstract{false};
+	ASTPointer<StorageLayoutSpecifier> m_storageLayoutSpecifier;
 
 	util::LazyInit<std::vector<std::pair<util::FixedHash<4>, FunctionTypePointer>>> m_interfaceFunctionList[2];
 	util::LazyInit<std::vector<EventDefinition const*>> m_interfaceEvents;
 	util::LazyInit<std::multimap<std::string, FunctionDefinition const*>> m_definedFunctionsByName;
 };
 
+
+class StorageLayoutSpecifier : public ASTNode
+{
+public:
+	StorageLayoutSpecifier(
+		int64_t _id,
+		SourceLocation const& _location,
+		ASTPointer<Expression> _expression
+	):
+		ASTNode(_id, _location),
+		m_expression(_expression)
+	{}
+	void accept(ASTVisitor& _visitor) override;
+	void accept(ASTConstVisitor& _visitor) const override;
+
+	ASTPointer<Expression const> expression() const { return m_expression; }
+private:
```

Comment:
No need to expose the smart pointer unless this is explicitly meant as a way to transfer (or share) ownership:

```suggestion
	Expression const& expression() const { return m_expression; }

private:
```

We don't generally do that with AST. The AST is the only thing that cares about ownership and we always keep it around as long as needed. The other parts of the compiler just get plain pointers (or even better - references).

Reply:
Same for `ContractDefinition::storageLayoutSpecifier()`. Also note that you're returning a non-`const` specifier, which allows it to be modified from the outside.

Reply:
> We don't generally do that with AST. The AST is the only thing that cares about ownership and we always keep it around as long as needed. The other parts of the compiler just get plain pointers (or even better - references).

Fixed.

> Also note that you're returning a non-`const` specifier, which allows it to be modified from the outside.

Fixed and introduced an overload which returns a non-const pointer because it is needed for `NameAndTypeResolver::resolveNamesAndTypesInternal`.

Reply:
Right, I see that `resolveNamesAndTypesInternal()` works only because many AST nodes already expose non-`const` stuff via `const` methods. E.g. `baseContracts()` or even `annotation()`. You should only be able to access these things if you have a mutable node.

But that's something we should fix separately.